feat(protocol-designer,-step-generation): build stacker step UI

ncdiehl11 opened the pull request:
# Overview

This PR fleshes out the flex stacker step form content. It also refactors some of our Redux infrastructure during container creations in deck setup to update the flex stacker's state.

Notable items to be addressed in a followup:
- correct rendering of labware added to the shuttle/hopper
- removal of stacker temporal property `maxPoolCount` and its implementations (should be derived directly from the hopper's set stored labware)

## Test Plan and Hands on Testing

- smoke test adding labware to the hopper and shuttle, and creating various hopper steps
- This is a big PR! It became untenable to split the UI work from the state updates to properly smoke test. I am happy to walk through live if helpful :) 

## Changelog

#### PD
- build full stacker step form UI, including several stacker state utilities
- update `createLabwareAboveModule` thunk and implementation in order to correctly set the stacker's state
- create and wire up new `updateStackerModuleState` action in Redux
- add new translations

#### step-generation
- update `stackerUpdates` logic to reflect decided-on state architecture
- refactor logic for determining safety of storing/retrieving/filling labware to/from the hopper

#### shared-data
- create and wire up correct logic for determining the max pool height given the stacker's set labware types

## Review requests

see test plan. reach out with any questions

## Risk assessment

low-medium. Large footprint, but mostly fully new features tied to the stacker only

ncdiehl11 committed:
feat(protocol-designer,-step-generation): build stacker step UI

This PR fleshes out the flex stacker step form content. It also refactors some of our Redux infrastructure during container creations in deck setup to update the flex stacker's state.

ncdiehl11 committed:
Merge branch 'edge' into feat_pd-stacker-step-ui

jerader commented on shared-data/js/helpers/getFlexStackerHardwareProps.ts:
```diff
@@ -76,3 +69,31 @@ export const getHeightOfLabwareStackFromDefinitions = (
+  const { labwareDefinitions, model } = args
+  const { primary, adapter, lid } = labwareDefinitions
+  const topDefinition = lid ?? adapter ?? primary
+  // bottom-up
```
oh god ya we need to audit all the hopper stacks to be bottom up now right?

ncdiehl11 replied:
yeppp

jerader commented on protocol-designer/src/step-forms/actions/thunks.ts:
```diff
@@ -31,21 +42,65 @@ export const createContainerAboveModule: (
+        )
+      }
+    }
+    console.log(getState())
```
```suggestion
```

jerader commented on protocol-designer/src/labware-ingred/actions/actions.ts:
```diff
@@ -79,6 +79,7 @@ export interface CreateContainerArgs {
   // NOTE: if slot is omitted, next available slot will be used.
   slot?: DeckSlot
   updateSelectedLabwareId?: boolean
+  uuids?: string[]
```
i might have missed this but why do we need to pass down a list of uuids from the args if in the `createContainer` thunk, we are making new uuids?

ncdiehl11 replied:
this is because we need access to the created uuids in the `createContainerAboveModule` thunk

jerader reviewed (commented):
woohoo! code looks good, just left a question about the uuid arg in `createContainer`. gonna smoke test now

ncdiehl11 committed:
fix(protocol-desiger): refactor module update dispatch to thunk

The responsibility of dispatching a module state update should be within the createContainerAboveModule thunk, not directly in the DeckSetup component. This allows us to correctly update stacker state for initial deck setup.

ncdiehl11 committed:
UI updates

ncdiehl11 committed:
resolve lint-css

ncdiehl11 committed:
fix CSS modules capitalization issue

ncdiehl11 committed:
fix tests, linting, js checks

ncdiehl11 committed:
Merge branch 'edge' into feat_pd-stacker-step-ui

jerader commented on protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepTools/FlexStackerTools/flexstackertools.module.css:
```diff
@@ -0,0 +1,61 @@
+.list_item {
```
oh ya good catch on the incorrect module.css file name before

jerader commented on protocol-designer/src/pages/Designer/ProtocolSteps/StepForm/StepTools/FlexStackerTools/index.tsx:
```diff
@@ -1,184 +1,212 @@
-        {t('protocol_steps:flex_stacker.stacker.label')}
-      </StyledText>
+  // preselect the first form option on mount if the form is presaved
+  useEffect(() => {
```
is there a way to remove this useEffect and instead populate it in the `createPresavedStepForm` fn or something? you don't have to fix this right now but would be nice to investigate it in a follow up

ncdiehl11 replied:
Definitely can investigate in a followup

jerader reviewed (commented):
will smoke test again when you fix the white screen we went over. This PR does a lot which is amazing! But can we make sure to document all the follow ups that need to be done? there are a few todo comments in here i think.

ncdiehl11 committed:
variety of step generation and PD form fixes to avoid white screen (handle unexpected null labware definitions passed to space-getting utilities)

ncdiehl11 committed:
lint

jerader reviewed (approved):
woohoo!!! Amazing work Nick! 🚀 🎉 was able to generate a protocol with a `retrieve()`, `store()`, and `empty()` python command

ncdiehl11 committed:
make sure labwareId is included in its stack after it is stored in a hopper
